app/controllers/internal_api/v1/pillars_controller.rb

```diff
@@ -166,5 +170,60 @@ def kubelet_contents
       }
     }
   end
+
+  def generate_user_block(con)
+    user = {}
+    user[:base_dn] = con.user_base_dn
+    user[:filter]  = con.user_filter
+    user_attr_map = user[:attr_map] = {}
+    user_attr_map[:username] = con.user_attr_username
+    user_attr_map[:id]       = con.user_attr_id
+    user_attr_map[:email]    = con.user_attr_email
+    user_attr_map[:name]     = con.user_attr_name
+    user
+  end
+
+  def generate_bind_block(con)
+    bind = {}
+    if con.bind_anon
+      bind[:anonymous] = true
+    else
+      bind[:anonymous] = false
+      bind[:dn] = con.bind_dn
+      bind[:pw] = con.bind_pw
+    end
+    bind
+  end
+
+  def generate_group_block(con)
+    group = {}
+    group[:base_dn] = con.group_base_dn
+    group[:filter] = con.group_filter
+    group_attr_map = group[:attr_map] = {}
+    group_attr_map[:user]  = con.group_attr_user
+    group_attr_map[:group] = con.group_attr_group
+    group_attr_map[:name]  = con.group_attr_name
+    group
+  end
+
+  def extra_dex_content
+    dex_connectors = DexConnectorLdap.all.map do |con|
+      {
+        "type"            => "ldap",
+        "id"              => con.id,
+        "name"            => con.name,
+
+        # Combine host and port since they ultimately feed into a single line of config for dex
+        "server"          => "#{con.host}:#{con.port}",
+        "start_tls"       => con.start_tls,
+        "root_ca_data"    => Base64.encode64(con.certificate.try(:certificate)),
```

Comment:
The UI portion that will be stacked on top of this should be created so as to make it unable to save a connector without having a certificate in place. That said, I agree it is a good change to make the rendering code resilient to the case anyway since it is possible in the database. I will make this so.

A bigger issue that is related that I see in the dev environment is that I have to manually create the table for the dex_ldap_connectors to bootstrap my dev cluster. My current understanding is that this is due to the way a static iso is used combined only partially with my local dev velum. I am not clear on this though, as it is weird to me that the table does not get automatically created in dev. I was told that changing the db models file and adding the migration db file should be sufficient when adding a new table. If there is something else that needs to be done I don't know what it is.

I cannot tell what is making the CI test fail now, as I don't know how to get the salt execution results out of a CI run. I suspect it is failing in the same way as my dev cluster; getting stuck because the DB table does not exist. I could handle this by adding guard code around the pillar output able to handle the table not existing, but it should be created. Handling the situation where the table doesn't exist seems like a hack.

Reply:
when you create a devenv from scratch, you'll see right at the beginning that a velum development container will be built. there we create the database as well with the new tables. so in theory that should work... but the thing is, you need to create it from scratch

the other solution would be to run the migrations inside the velum container, and then commit that container so the change persists in your env, but i haven't done that yet as it's easier to create all from scratch

CI is green now :+1: